### PR TITLE
Add an apollo-fetch-upload package

### PR DESCRIPTION
**Pull request by jaydenseric**
An `apollo-fetch-upload` package to fix [#6](https://github.com/apollographql/apollo-fetch/issues/6), superseding [#8](https://github.com/apollographql/apollo-fetch/pull/8).

This package allows the use of files anywhere in a GraphQL request (in theory, not just in `variables`) and is [compatible](https://github.com/apollographql/apollo-fetch/issues/6#issuecomment-314039458) with [`apollo-upload-server`](https://github.com/jaydenseric/apollo-upload-server).

To try it out:

1. Clone this PR.
2. Within `apollo-fetch/packages/apollo-upload-fetch` run `npm install && npm link`.
3. Clone the [apollo-upload-examples `apollo-fetch-upload` branch](https://github.com/jaydenseric/apollo-upload-examples/tree/apollo-fetch-upload).
4. Within `apollo-upload-examples/app` run `npm install && npm link apollo-upload-fetch`. You may have to manually remove `apollo-upload-fetch` from the dependencies first since it is not published yet.
5. Following the remain setup instructions for `apollo-upload-examples`.

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [ ] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

### CODE REVIEW

**Comment by jaydenseric** [comment]
I haven't testing it with batching yet, but it should work.

**Comment by jaydenseric** [comment]
There is a `FetchOptions` type I would like to use, but [#13](https://github.com/apollographql/apollo-fetch/pull/13) needs to be merged first so it can be imported from `apollo-fetch`.

File: packages/apollo-fetch-upload/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  ApolloFetch,
+  GraphQLRequest,
+} from 'apollo-fetch';
+import { extractFiles } from 'extract-files';
+
+export function constructUploadOptions(
+  requestOrRequests: GraphQLRequest | GraphQLRequest[],
+  options: RequestInit,
+): RequestInit {
+  const files = extractFiles(requestOrRequests);
```
Comment (jaydenseric) [inline packages/apollo-fetch-upload/src/index.ts]: @evanshauser Does it matter that `requestOrRequests` is modified by `extractFiles`? Or would it be better if extract files did not modify the input:

```
const { files, modifiedTree } = extractFiles(requestOrRequests)
```

Keep in mind that there are performance considerations to messing around with objects that contain potentially megabytes of files.

Comment (evanshauser) [inline packages/apollo-fetch-upload/src/index.ts]: @jaydenseric currently, `reqeustOrRequests` isn't used after it is incorporated into the fetch options, so should be safe to modify in `constructOptions`/we don't need a defensive copy in `apollo-fetch`.

In general from `apollo-fetch`'s perspective, once we have called an external function, such as `constructOptions`, the references passed as arguments should not be used by `apollo-fetch`

**Comment by jaydenseric** [comment]
I think the implementation is ready. @evanshauser are you able to explain how to get batching to work [here](https://github.com/jaydenseric/apollo-upload-examples/blob/apollo-fetch-upload/app/helpers/init-apollo-client.js#L20) so we can test it out?

Help setting up the tests will be much appreciated, I'm not really sure how to go about it. `extract-files` has it's own tests, so we just need to test the APIs defined in this package.

**Comment by evanshauser** [comment]
@jaydenseric awesome work!

The way to get batching enabled is ([batch blog](https://dev-blog.apollodata.com/query-batching-in-apollo-63acfd859862)):
```js
const client = new ApolloClient({
 // ... other options ...
 shouldBatch: true,
});
```

This option would mean that all queries would be put into a single batch, so file uploads and regular GraphQL requests would be placed in the same network request. Does your `apollo-fetch-upload` differentiate/handle the two during `constructOptions`?

Also, can `FormData` encode regular queries? If it's not compatible with the default `stringify` request, then one idea is to a function for `apollo-link`'s [split](https://github.com/apollographql/apollo-link/blob/master/docs/implementation.md#split-points) that routes file uploads to one Link and the rest to another. Something like:

screen shot 2017-07-27 at 2 55 01 pm

We're working on a BatchingLink that will perform the batching logic and run a provided function on the batch.

**Comment by jaydenseric** [comment]
@evanshauser I'm quite familiar with batching, and `shouldBatch` is deprecated. The [current way](http://dev.apollodata.com/core/network.html#BatchingExample) is to use `batchingNetworkInterface`:

```js
import ApolloClient, { createBatchingNetworkInterface } from 'apollo-client'

const client = new ApolloClient({
  networkInterface: createBatchingNetworkInterface({
    uri: '/graphql'
  })
})
```

But with `apollo-fetch`, `networkInterface` is replaced with:

```
networkInterface: { query: req => apolloFetchUpload({ ...req, query: print(req.query) }) }
```

`apollo-fetch-upload` works the same with regular or batched requests. If a request or batch of requests contains no files at all, the request is sent as default, without `FormData`.

**Comment by stubailo** [comment]
Yeah - right now all of the logic for batching (collecting multiple requests and sending them as one) lives inside the network interface implementation. I think the correct solution here is to use `apollo-link` to do the batching, which will pass down an array of requests down to `apollo-fetch`. I think @evanshauser is working on a batching link as we speak.

**Comment by evanshauser** [comment]
@jaydenseric This looks great! I think we are ready to merge after unit tests for getting the transformation tested for a single request(upload and normal with/without an array) and batched request(two normal, two uploads, and one of each). Not sure about the different errors that can occur from a failed upload, so I'll trust you to include what you feel is relevant. Then after the changelogs are per project, we can announce apollo-fetch-upload!

So excited, this will be an awesome addition!

We're also planning to add some Continuous deployment into Travis-CI, so if you could add the npm user, mdg, to the apollo-fetch-upload, we can start to hook things up.

**Comment by jaydenseric** [comment]
@evanshauser So, it's been an unproductive day attempting to create tests. Here are some of the challenges...

`File` and `FormData` APIs are not available in Node.js. It seems that there is very little info on unit testing browser code that uses these APIs in Node.js. It's an unsolved problem?

I attempted to polyfill `FormData` using packages such as [`form-data`](https://github.com/form-data/form-data) but appending files is different to web standards; they don't understand `File` instances. I could not find a suitable polyfill for `File` and even if I did, the `FormData` polyfill would not play nice.

For these reasons there are no tests in [`apollo-upload-client`](https://github.com/jaydenseric/apollo-upload-client). Maybe we should publish a v1 without all the tests we would like to have. For now `apollo-fetch-upload` is opt-in so I don't think its unreasonable. [`extract-files`](https://github.com/jaydenseric/extract-files) has tests, and that's where the most confusing bugs originate.

**Comment by HriBB** [comment]
How do [supertest](https://github.com/visionmedia/supertest) and [superagent](https://github.com/visionmedia/superagent) do it?

In this case I am starting a `Koa` server, then sending a `multipart/form-data` request to the `/graphql` endpoint. Maybe you could pass this data directly to the test? 
```js
import Koa from 'koa'
import body from 'koa-bodyparser'
import { resolve } from 'path'
import request from 'supertest'
import { expect } from 'chai'
import 'mocha'

const app = new Koa()

app.use(body())

createGraphQLServer(app)

const server = app.listen(4004)

describe('GraphQL Server', () => {

  it('can upload images', () => {
    return request(server)
      .post('/graphql')
      .field('operationName', 'uploadImages')
      .field('query', `mutation uploadImages($location_id: Int!, $images: [File!]!) {
        uploadImages(location_id: $location_id, images: $images) {
          name slug width height file_name extension mime_type url
          owner { id username }
        }
      }`)
      .field('variables', JSON.stringify({
        location_id: 3,
      }))
      .attach('images', resolve(__dirname, 'images', 'image1.jpg'))
      .attach('images', resolve(__dirname, 'images', 'image2.jpg'))
      .attach('images', resolve(__dirname, 'images', 'image3.jpg'))
      .expect(200)
      .then(res => {
        expect(res.body).to.deep.equal(expected)
      })
  })

})
```
Probably I'm missing something ...

**Comment by evanshauser** [comment]
@jaydenseric This is incredible work! Great job! So happy to have `apollo-fetch-upload`

**Pull request merged**